### PR TITLE
[DRAFT]Mistral-Small-3.1-24B Model support on Blackhole QB-2

### PR DESCRIPTION
**Pull request by ign-steve**
## Summary

Migrated Mistral-Small-3.1-24B-Instruct vision–text inference to Blackhole QuietBox-2 (4× P150, mesh 1×4) with prefill+decode trace, including a multimodal-aware prefill trace path. Wormhole support is unchanged.

### Blackhole / mesh enablement
- Added `P150x4: (1, 4)` to the `mesh_device` parameterization in `tests/pipeline_tests/test_end2end.py` so the E2E test runs on BH QB-2.
- Increased `trace_region_size` from `30 MB → 35 MB` across the pipeline/vision tests — required to fit prefill+decode trace buffers on P150×4 (`mesh_trace` allocator limit).
- Raised `num_command_queues` from `1 → 2` in the E2E test to overlap decode trace submit with token readback.

### Multimodal prefill trace
- New `MistralGenerator` (`tt/generator.py`) and prefill helpers on `MistralTransformer` add a multimodal-aware trace path: trace input is **text-only** `uint32` ids or **pre-fused `bfloat16`** vision+text embeddings, cached on capture and reused on replay to skip the redundant vision forward.

- `prepare_inputs_prefill` now uses padded `cos_matrix_prefill`/`sin_matrix_prefill` slices so trace capture always sees a `max_seq_len`-shaped rope input.

### Vision-tower / projector optimizations
- `tt/pipeline/mistral_vision_tower.py`: added a single-full-image fast path that skips the `transpose → reshape → slice → reshape → transpose` chain (removes 2× `ReshapeView` + 2× `transpose` from the prefill trace). Replaced `position_ids_in_meshgrid_tt` (ttnn meshgrid + host round-trip) with a small `_meshgrid_position_ids_torch` helper.

- `tt/vision_mmp.py::TTMistral3PatchMerger`: replaced the host `torch.nn.functional.unfold` round-trip with an on-device `reshape → permute → reshape` matching unfold ordering; eliminates a `to_torch`/`from_torch` per image. Set `TTMistral3MultiModalProjector.norm` to `mode="prefill"` to match the surrounding prefill execution.

### Decode Loop (`test_end2end.py`)
- Decode loop moved to on-device greedy sampling with non-blocking traced replay (`enable_trace=True`, `read_from_device=False`);

- Dropped the host `sample_host` / tokenizer-decode hot path and added a perf-measurement logger (TTFT, prefill/decode t/s, full runtime).

### Test cleanups
- Vision unit tests (`test_vision_model`, `test_vision_tower`, `test_pixtral_transformer`, `test_vision_attention`, `test_vision_mlp`, `test_vision_rms`, `test_patch_rot_emb`) drop `.float()` casts on reference inputs so torch and TTNN run at matching `bfloat16` precision.

---
## Performance Measurements

End-to-end demo metrics 

| Metric                                  | Without trace    | With trace |
| --------------------------------------- | ------------- | ------------- |
| **prefill_tokens**                      | 522           | 522           |
| **inference_prefill** (s)               | 2.82139       | 3.29152       |
| **prefill_t/s** (tokens/s)              | 185.015       | 158.59        |
| **TTFT** (prefill to first token, ms)   | 2821.39       | 3291.52       |
| **inference_decode** (s)               | 27.0651       | 13.9218       |
| **decode_t/s/u** (per-user)             | 17.4025       | 33.832        |
| **decode_t/s** (aggregate, batch=1)     | 17.4025       | 33.832        |
| **num_decode_tokens**                   | 471           | 471           |
| **Full demo runtime** (s)               | 29.9914       | 17.4228       |

<small>Note: The above performance numbers are from runs after optimization.</small>

| Metric                                | Non-Optimized | Optimized |
| ------------------------------------- | ---------------------- | ----------------------------- |
| **prefill_tokens**                    | 522                    | 522                           |
| **inference_prefill** (s)            | 3.50515                | 3.29152                       |
| **prefill_t/s** (tokens/s)            | 148.924                | 158.59                        |
| **TTFT** (prefill to first token, ms) | 3505.15                | 3291.52                       |
| **inference_decode** (s)              | 15.0146                | 13.9218                       |
| **decode_t/s/u** (per-user)           | 29.3047                | 33.832                        |
| **decode_t/s** (aggregate, batch=1)   | 29.3047                | 33.832                        |
| **num_decode_tokens**                 | 440                    | 471                           |
| **Full demo runtime** (s)             | 18.8109                | 17.4228                       |

<small>Note: Both optimized and non-optimized performance numbers are from runs with prefill and decode trace enabled.</small>

---

### End-to-end pipeline

```bash
export HF_MODEL=mistralai/Mistral-Small-3.1-24B-Instruct-2503
pytest models/experimental/mistral_24b/tests/pipeline_tests/test_end2end.py
```

- Added ISL sweep results and decoder logit pcc values in the README. 

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/mistral_blackhole_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/mistral_blackhole_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ign/mistral_blackhole_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/mistral_blackhole_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/mistral_blackhole_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/mistral_blackhole_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/mistral_blackhole_support)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/mistral_blackhole_support)

### CODE REVIEW

**Comment by moritztng** [comment]
Optimization knobs are in good shape.

**But this is required:**

- **Add a text decoder PCC test.** Vision components have PCC tests at 0.99, but the text decoder has no logit-level correctness validation. The E2E test only checks that tokens are generated, not that they are correct. DOTS-OCR has `test_text_prefill_pcc.py` comparing TT vs HF logits at PCC >= 0.98. Add an equivalent.

**Comment by ign-steve** [comment]
> Optimization knobs are in good shape.
> 
> **But this is required:**
> 
> * **Add a text decoder PCC test.** Vision components have PCC tests at 0.99, but the text decoder has no logit-level correctness validation. The E2E test only checks that tokens are generated, not that they are correct. DOTS-OCR has `test_text_prefill_pcc.py` comparing TT vs HF logits at PCC >= 0.98. Add an equivalent.

Have added a text-decoder PCC test comparing TT vs HF logits at PCC >= 0.99
